src/projections.jl

```diff
@@ -295,6 +396,61 @@ function projection_gradient_on_set(::DefaultDistance, v::AbstractVector{T}, ::M
     return D
 end
 
+"""
+    projection_gradient_on_set(::DefaultDistance, v::AbstractVector{T}, ::MOI.ExponentialCone) where {T}
+
+derivative of projection of vector `v` on closure of the exponential cone,
+i.e. `cl(Kexp) = {(x,y,z) | y e^(x/y) <= z, y>0 } U {(x,y,z)| x <= 0, y = 0, z >= 0}`.
+
+References:
+* [Solution Refinement at Regular Points of Conic Problems](https://stanford.edu/~boyd/papers/cone_prog_refine.html)
+by Enzo Busseti, Walaa M. Moursi, and Stephen Boyd
+"""
+function projection_gradient_on_set(::DefaultDistance, v::AbstractVector{T}, s::MOI.ExponentialCone) where {T}
+    _check_dimension(v, s)
+    Ip(z) = z >= 0 ? 1 : 0
+
+    if distance_to_set(DefaultDistance(), v, MOI.ExponentialCone()) < 1e-8
+        return Matrix{Float64}(I, 3, 3)
+    end
+    if distance_to_set(DefaultDistance(), -v, MOI.DualExponentialCone()) < 1e-8
+        # if in polar cone Ko = -K*
+        return zeros(3,3) #FillArrays.Zeros(3, 3)
+    end
+    if v[1] <= 0 && v[2] <= 0
+        return Diagonal([1; Ip(v[2]); Ip(v[3])])
```

Comment:
Sorry I realize now I misled you here.
Ideally, we want to always return the same type from a function (so that the caller knows what it is getting).
So to be consistent, we should return a normal Matrix here, so that all the possible return values of the function have the same type

```suggestion
        return LinearAlgebra.diagm(0 => T[1, Ip(v[2]), Ip(v[3])])
```

Reply:
also I converted to the element type of the vector